### DISCUSSION TITLE
[Q&A] Glpi_agent - network Discovery/inventory on windows/Linux agent

### DISCUSSION THREAD

**Discussion by Chico0008** [glpi-project/glpi-agent#317] [Q&A]
### Documentation & bug reporting acknowledgment

Yes, I read it

### Describe your problem

Hi
i'm using a windows 10 computer with the last nightly of GLPI agent (1.5-git05880ee3)
eveything regarding my computer is fully inserted in my Glpi Inventory (using Glpi 10.0.5 and native inventory on a linux Ubuntu server)

But, can't find a way to use the agent, or server, to run a network discovery/inventory, to get switch/firewall/etc (passive equipement) to get into my inventory

thanks

**Comment by g-bougard** [comment]
Hi @Chico0008 
you first need to configure your agent with the right URL, [read cautions regarding server parameter in online documentation](https://glpi-agent.readthedocs.io/en/latest/configuration.html#server).
Then in glpiinventory plugin, you have to create ip range and associate snmp credentials to it.
Then you can create a netdiscovery task in glpiinventory for which you'll have to select an agent actor and your iprange as target. Then you can click on the "force a start" button on the general task tab.
Here, the next time the agent actor will contact the server it will be notified it has a netdiscovery task to run.

In the same way, you'll have to create a netinventory task which always have to be started after the netdiscovery one as it will make a deeper inventory on devices which only supports SNMP.

In GLPI, you also must check the agent modules netdiscovery & netinventory are enabled for the agent actor.

In the agent side, the tasks must have been installed.

You can force manually the agent to run its tasks, handy during testing.

**Comment by Chico0008** [comment]
I'm using native inventory in glpi, not a plugin, can the native inventory handle this ?
or do i must install the Glpi Inventory plugin ?

**Reply by g-bougard** [reply]
Actually GlpiInventory plugin is required if you want to manage tasks like netdiscovery or netinventory from GLPI.
This tasks support will be added natively in GLPI in the future.

**Comment by Chico0008** [comment]
I activate toolbox on my agent, but it says things are not installed.
![Capture](https://user-images.githubusercontent.com/58886243/213156113-49987c4b-ff04-4f25-848b-2a96f28df8f2.PNG)
how can i install theses 2 plugins on my windows agent ?

**Reply by g-bougard** [reply]
During glpi-agent installation, you must select network tasks.

**Reply by Chico0008** [reply]
i did, used this parameter for installation :
ADDLOCAL=feat_DEPLOY,feat_COLLECT,feat_NETINV

but in local toolbox, still saying netdiscovery and netinventory not installed :(

**Reply by g-bougard** [reply]
Something goes wrong on your side. I just tried a quiet installation with your parameter with the same nightly build and I don't have this message in ToolBox.
![image](https://user-images.githubusercontent.com/12514489/213185726-e7e2675a-aba9-4dde-b7b0-58e0886ceeb2.png)

Can you open an admin console, `cd` to the GLPI-Agent installation folder and run the following command to list installed tasks ?
```
glpi-agent --list-tasks
```

**Reply by Chico0008** [reply]
```
C:\Windows\system32>cd /d C:\Program Files\GLPI-Agent

C:\Program Files\GLPI-Agent>glpi-agent --list-tasks

Available tasks :
- RemoteInventory (v1.0)
- Collect (v2.8)
- Inventory (v1.12)
- Deploy (v2.10)

target server0: server http://glpi-test.viry.net/front/inventory.php
Planned tasks: Inventory,Collect,RemoteInventory,Deploy

C:\Program Files\GLPI-Agent>C
```

**Reply by g-bougard** [reply]
This shows the tasks are not installed.

Did you eventually tried to reinstall other an existing installation on which the tasks was not installed ?
With silent installation, this just does not work if you're re-installaing the same version and you're not advertised of the problem (probably only in system eventviewer).
But in that case, you can try to add the following parameter to force reinstallation: `REINSTALL=feat_AGENT`

**Reply by Chico0008** [reply]
Juste uninstalled the agent via windows parameter, and installed again, now its OK, i have the network options.
but now
i installed Glpi Inventory, set IP range, create a task for discovering network, set the range, set my PC as agent.
save, force run, and ... nothing.

in task supervision, i have task in to do, but nothing is running :s
on my agent, on toolbox, iprange is not displayed.
![Capture](https://user-images.githubusercontent.com/58886243/213191725-c31dad6a-1918-4b8e-b35b-4fd61242da30.PNG)

**Reply by Chico0008** [reply]
Seems OK now, task are running.
thanks for your help

**Reply by g-bougard** [reply]
Nice.
Just for you to know, ToolBox and GlpiInventory interface are not linked. You can use ToolBox without GlpiInventory. In ToolBox, edit the configuration clicking on the gear on the upper right corner and enable the menus for IPRange & Credentials. Then you will be able to create them in ToolBox too.
Actually, ToolBox can only be used to run manually network scan, storing XML locally. There's no scheduling nor automatic submission to GLPI. But you still can easily inject the stored XML in GLPI with glpi-injector script.
P.S.: I'm actually working on a ToolBox update to be able to schedule network scans.

**Reply by Chico0008** [reply]
Ok, thanks, i thought they were linked.

**Comment by Chico0008** [comment]
So, i wanted to do things right and make a task running at night for discovery / enventoring, using a Linux agent on the Ubuntu server hosting Glpi.
my agent on linux can do everything, i plan my task to run yesterday a 8pm, but nothing happens.
nothing in status, not prepaped, not ready, nothing :(

the linux agent is supposed to contact the server around every 2 hours, and i can confirm it's right because his inventory last update is ok.

**Reply by g-bougard** [reply]
Enable debug=2 in conf, restart the service and check the agent log. On linux, if you don't log in a log file, you can use the following command to follow the agent log in a console.
```
journalctl -f -t glpi-agent
```
You can remove the `-f` option to consult all the log from the beginning.
This should help you to understand what's going on.

**Comment by Chico0008** [comment]
Debug=2 enable, agent restarted
agent is supposed to contact the server at 9;37
![1](https://user-images.githubusercontent.com/58886243/213650345-996f350d-0b40-4fc8-b9c8-3a4879be4006.PNG)
I planned a network discovery at 9:40
![2](https://user-images.githubusercontent.com/58886243/213650367-875162d3-3fc8-4c4f-81fb-c34933d307c2.PNG)

i'll get you the log after that

edit : Here's the log, still no network discovey running :(

```
[Fri Jan 20 09:37:38 2023][info] target server0: server http://glpi-test.viry.net/front/inventory.php
[Fri Jan 20 09:37:38 2023][debug2] [http client] Using Compress::Zlib for compression
[Fri Jan 20 09:37:38 2023][info] sending contact request to server0
[Fri Jan 20 09:37:38 2023][debug2] [http client] C1072423: sending message:
...
lots of server local inventory, vg, package, etc
...
[Fri Jan 20 09:38:24 2023][debug2] format: Zlib
[Fri Jan 20 09:38:24 2023][debug2] [http client] 2B801E85: received message:
{"expiration":"2","status":"ok"}
[Fri Jan 20 09:38:24 2023][debug] forking process 6897 to handle child job
[Fri Jan 20 09:38:24 2023][debug2] child[6897]: forked
[Fri Jan 20 09:38:24 2023][debug] forking process 6897 to handle task WakeOnLan
[Fri Jan 20 09:38:24 2023][debug2] child[6783] finished
[Fri Jan 20 09:38:24 2023][debug] WakeOnLan task execution not requested
[Fri Jan 20 09:38:24 2023][debug] forking process 6900 to handle child job
[Fri Jan 20 09:38:24 2023][debug2] child[6900]: forked
[Fri Jan 20 09:38:24 2023][debug] forking process 6900 to handle task Collect
[Fri Jan 20 09:38:24 2023][debug2] child[6897] finished
[Fri Jan 20 09:38:24 2023][info] running task Collect
[Fri Jan 20 09:38:24 2023][debug2] [http client] Not using compression
[Fri Jan 20 09:38:24 2023][debug2] http://glpi-test.viry.net/front/inventory.php?action=getConfig&task[Collect]=2.8&machineid=s-web-2023-01-19-14-00-46
[Fri Jan 20 09:38:24 2023][error] [http client] communication error: 400 Bad Request, Protocol not supported
[Fri Jan 20 09:38:24 2023][info] Collect task not supported by server0
[Fri Jan 20 09:38:24 2023][info] target server0: next run: Fri Jan 20 11:36:53 2023 - http://glpi-test.viry.net/front/inventory.php
[Fri Jan 20 09:38:24 2023][debug2] child[6900] finished
```

not Discovery planned, the agent is able to do it, result of glpi-agent --list-tasks
```
Available tasks :
- RemoteInventory (v1.0)
- NetDiscovery (v5.1)
- Collect (v2.8)
- Inventory (v1.12)
- NetInventory (v5.1)
- ESX (v2.7)
- Deploy (v2.10)
- WakeOnLan (v2.2)

target server0: server http://glpi-test.viry.net/front/inventory.php
Planned tasks: Collect,Inventory,RemoteInventory,NetDiscovery,Deploy,ESX,WakeOnLan,NetInventory
```

**Reply by Chico0008** [reply]
Still nothing Friday afternoon, i had to do a : `glpi-agent --task=netdiscovery` on the linux server, so the task could be taken, and finally showed to the state Prepared, and waiting on Glpi server.
so i guess i had to cron netdiscovery et netinventory on linux ?

**Reply by g-bougard** [reply]
Hi @Chico0008 
```
[Fri Jan 20 09:38:24 2023][debug2] http://glpi-test.viry.net/front/inventory.php?action=getConfig&task[Collect]=2.8&machineid=s-web-2023-01-19-14-00-46
```
go back to my first answer where I was telling you to read carefully some Cautions in doc... Your server URL in agent configuration is just wrong. RFM again ;-)

**Reply by Chico0008** [reply]
?? i'm using this url and it work for inventory ????
i used this one, because on the root (http://glpi-test.viry.net) i have a kerberos/SSO login process

if i set the agent to use root url, it won't work because of this
in the manual it says the ./front/inventory.php can be used, so i use it, and in my apache configuration this url is set to not use SSO.

if i must use root url for everything, i don't know what apache configuration i have to use to keep SSO for regular users, and other process for agent.

in manual
```
If you're not using GlpiInventory plugin:
Your server URL should look like: https://my-glpi-server/, or https://my-glpi-server/front/inventory.php
```

**Reply by g-bougard** [reply]
You're probably reading too quickly the doc, `/front/inventory.php` is to be used when you only need native inventory. Since you're using GlpiInventory for NetDiscovery task, you need to use `/marketplace/glpiinventory/` or `/plugins/glpiinventory`.

**Reply by Chico0008** [reply]
![image](https://user-images.githubusercontent.com/58886243/214022538-7ad2f6b7-2200-4bda-8a70-a3f803a70022.png)

sooooo
just reinstalled my linux agent, with all modules, pointing to the correct url (/plugins/glpiiinventory)
first contact and local inventory are fine

set up my networdk discovery task to run at 12:00:00, i'll see in the afternoon if it has run.

**Reply by Chico0008** [reply]
![image](https://user-images.githubusercontent.com/58886243/214048956-0da14863-67f8-4d0e-9496-7af98d958bb4.png)

tries Inventory for printer, seems OK
just 1 or 2 not getting any informations, dk why, but it's fine

last question, Tasks are run every day ?
because i can set a Start date/hour, but then ? how do i make them run everyday ?

**Reply by g-bougard** [reply]
Tasks should be "prepared" before the agent contacts the server. You can use scheduling on task and the "taskscheduler" automatic action if enabled (check it) so the task is always ready to be run when the agent will run tasks next time. The "execution" scheduling is related to "wakeupAgents" automatic action and will try to force agent to run tasks (including inventory) during the selected period.

**Reply by Chico0008** [reply]
Not sure i understood right
here's one of my task
![Capture](https://user-images.githubusercontent.com/58886243/214060928-b39cf337-eca2-47ed-84e3-ebc97d6a3a00.PNG)
it should start tonight at 10pm
but after that, will it run again tomorrow a 10pm ?

task is set to run on the agent on the linux server, no need to wakeonlan, and the agent contact the server every 2 hours, meaning the task could run near midnight at most ?
(agent contact server a 9:59, no run time, next contact a 11:59, task was at 10, so run ?)

**Reply by Chico0008** [reply]
Ok i'm really going to reap my hairs from my head
i planned my task for last night
![1](https://user-images.githubusercontent.com/58886243/214237792-dd12fac2-80ac-4074-ad44-8b8d84f7e0a8.PNG)
and nothing has been done

i really don't get it
the last time it work, i force the execution and ran a glpi-agent on the server
but last night, nothin, wasn't it supposed to do all alone ?

sorry but i really don't get how this is supposed to work.

**Reply by g-bougard** [reply]
You must check if the task in GlpiInventory is in the "prepared" state. If not, check the "taskscheduler" automatic action history in GLPI Configuration.
If tasks are still in the "prepared" state. Check the glpi-agent log, did it run at the time you expected ?
You know, telling this doesn't work without sharing relevant information doesn't help us to understand what you or we missed.

**Reply by Chico0008** [reply]
tried to forcerun this morning, but task even won't get in prepared state
how can i get them in this state automaticaly ?

**Reply by g-bougard** [reply]
If a task remains not prepared when you click on the "force a start" button, this probably means something goes wrong in the task setup, maybe the agent is no more associated to the task.

**Comment by Chico0008** [comment]
Agent is still associated in job setup.
tried to delete, add again, save, force

this time 2 jobs or prepared (the first 2 on the list)
but the 3 others, even doing this don't go to prepared state :/

**Reply by Chico0008** [reply]
i'll use google trad because my english begin to be bad. (french origin)

Well, yesterday by doing forced executions, it worked.

I wanted to plan a printer inventory task yesterday at 8 p.m.
I put a preparation slot from 6 pm to 8 pm, and an execution slot from 12 am to 7 am

this morning, still nothing, the task has not even been prepared
in the agent's log, nothing (just its local inventory)

I can't find a doc explaining the daily planning of a task, or else I have a problem somewhere, but I can't find where it can block.

**Reply by g-bougard** [reply]
Have you checked "taskscheduler" automatic action was run in GLPI Configuration ?
![image](https://user-images.githubusercontent.com/12514489/214512561-fe3e1449-311c-47bc-aaf1-6d35977c1ec3.png)

As mine is on a test server and I didn't setup crontab to automatically run automatic actions, the "Last run" date is indeed the last time I manually run it, clicking on the "Execute" button:
![image](https://user-images.githubusercontent.com/12514489/214513283-e1ac5469-f091-4254-bff1-a59fb9ef01b2.png)

The log should tell when it was run:
![image](https://user-images.githubusercontent.com/12514489/214513639-d682168e-4e12-4b2d-a75b-4aa2d9ae868f.png)

**Reply by Chico0008** [reply]
nothing for me
![image](https://user-images.githubusercontent.com/58886243/214516535-213d415e-301e-49c4-8761-047d0796d9d2.png)

it's this taskscheduler, making inventory task to be run ?

**Reply by g-bougard** [reply]
As I said 2 times yesterday ;-), this automatic action is here to "prepare" the task.
Check you have configured "Automatic actions" in CLI mode from crontab, see GLPI documentation:
https://glpi-user-documentation.readthedocs.io/fr/master/modules/configuration/crontasks.html

**Reply by Chico0008** [reply]
you said task needed to be prepared, but i didn't know this was made by another task in glpi, not easy to understand all the things linked in glpi backend.
So i'll add a cron on my minux to launch the cron.php.
